### ISSUE TITLE
可以提供一下图片推断可视化的代码吗

### ISSUE THREAD

**Issue by NYB-AAA** [xiaoqiang-lu/LSST#1]
源码里似乎只有训练和测试代码，没有图片推断代码，可以提供一下吗，感谢

**Comment by xiaoqiang-lu** [comment]
你好，将此函数放置在测试代码中调用即可：

    def eval_visual(model, valloader, args):

      model.eval()
      tbar = tqdm(valloader)
  
      cmap = color_map(args.dataset)
  
      with torch.no_grad():
          for img, _, id in tbar:
              img = img.cuda()
              pred = model(img)
              pred = torch.argmax(pred, dim=1).cpu()
  
              pred = Image.fromarray(pred.squeeze(0).numpy().astype(np.uint8), mode='P')
              pred.putpalette(cmap)
              pred.save('%s/%s' % (args.save_path, os.path.basename(id[0].split(' ')[1])))

**Comment by NYB-AAA** [comment]
感谢！



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年11月8日(周三) 晚上11:14
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [xiaoqiang-lu/LSST] 可以提供一下图片推断可视化的代码吗 (Issue #1)




 
你好，将此函数放置在测试代码中调用即可：
 def eval_visual(model, valloader, args):   model.eval()   tbar = tqdm(valloader)   cmap = color_map(args.dataset)   with torch.no_grad():       for img, _, id in tbar:           img = img.cuda()           pred = model(img)           pred = torch.argmax(pred, dim=1).cpu()           pred = Image.fromarray(pred.squeeze(0).numpy().astype(np.uint8), mode='P')           pred.putpalette(cmap)           pred.save('%s/%s' % (args.save_path, os.path.basename(id[0].split(' ')[1])))

**Comment by NYB-AAA** [comment]
您好，我用提供的函数直接调用会报错：
![image](https://github.com/xiaoqiang-lu/LSST/assets/130569364/f790852b-ce1f-427f-ab27-e3e208487a78)

为了解决报错，我在pred = Image.fromarray(pred.squeeze(0).numpy().astype(np.uint8), mode='P')前加入 pred = pred[0]，如下：

```
def eval_visual(model, valloader, args):
    model.eval()
    tbar = tqdm(valloader)

    cmap = color_map(args.dataset)

    with torch.no_grad():
        for img, _, id in tbar:
            img = img.cuda()
            pred = model(img)
            pred = torch.argmax(pred, dim=1).cpu()
            pred = pred[0]
            # pred = pred.permute(1, 2, 0)
            # pred = Image.fromarray(pred.numpy().astype(np.uint8), mode='RGB')
            pred = Image.fromarray(pred.squeeze(0).numpy().astype(np.uint8), mode='P')

``
            # pred = Image.fromarray(pred.squeeze(0).numpy().astype(np.uint8), mode='RGB')
            pred.putpalette(cmap)
            pred.save('%s/%s' % (args.save_path, os.path.basename(id[0].split(' `')[1])))`
```
但是这样做之后得到的图片都是黑色的，不是论文里的彩色分割结果，请问如何解决这个问题呢

**Comment by xiaoqiang-lu** [comment]
不好意思，我的错误。下述代码可正常工作。

    def eval_visual(model, valloader, args):

        model.eval()
        tbar = tqdm(valloader)
    
        cmap = color_map(args.dataset)
    
        with torch.no_grad():
            for img, _, id in tbar:
                img = img.cuda()
                pred = model(img)
                pred = torch.argmax(pred, dim=1).cpu()
                for i in range(pred.shape[0]):
                    output = Image.fromarray(pred[i].squeeze(0).numpy().astype(np.uint8), mode='P')
                    output.putpalette(cmap)
                    output.save('%s/%s' % (args.save_path, os.path.basename(id[i].split(' ')[1].split('/')[1])))

**Comment by NYB-AAA** [comment]
经过测试，这个代码确实可以将图片推断可视化，代码训练和测试得到的miou都基本符合论文里的值，但是为什么图片可视化结果是黑色的不是彩色的分割结果呢，求解答




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "xiaoqiang-lu/LSST"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2023年11月10日(星期五) 下午4:18
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [xiaoqiang-lu/LSST] 可以提供一下图片推断可视化的代码吗 (Issue #1)





 
不好意思，我的错误。下述代码可正常工作。
 def eval_visual(model, valloader, args):     model.eval()     tbar = tqdm(valloader)     cmap = color_map(args.dataset)     with torch.no_grad():         for img, _, id in tbar:             img = img.cuda()             pred = model(img)             pred = torch.argmax(pred, dim=1).cpu()             for i in range(pred.shape[0]):                 output = Image.fromarray(pred[i].squeeze(0).numpy().astype(np.uint8), mode='P')                 output.putpalette(cmap)                 output.save('%s/%s' % (args.save_path, os.path.basename(id[i].split(' ')[1].split('/')[1])))

**Comment by xiaoqiang-lu** [comment]
不好意思，这是由于代码中的color_map函数没有对应到正确的数据集名称。请将utils.py中的第122行`elif dataset == 'MF_DFC22':`更改成`elif dataset == 'DFC22':`即可。保证color_map函数中所需的数据集名称和你所使用的数据集名称相同就行。

**Comment by NYB-AAA** [comment]
感谢感谢，我明早去试试



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年11月10日(周五) 晚上9:57
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [xiaoqiang-lu/LSST] 可以提供一下图片推断可视化的代码吗 (Issue #1)




 
不好意思，这是由于代码中的color_map函数没有对应到正确的数据集名称。请将utils.py中的第122行elif dataset == 'MF_DFC22':更改成elif dataset == 'DFC22':即可。保证color_map函数中所需的数据集名称和你所使用的数据集名称相同就行。

**Comment by NYB-AAA** [comment]
非常感谢，修改完之后可以可视化了，请问可以提供一下预处理的代码吗，想看一下是如何对数据集进行处理的

**Comment by xiaoqiang-lu** [comment]
具体对于数据集的处理代码在`dataset/semi.py`，关于数据增强处理部分在`dataset/transform.py`处。

**Comment by NYB-AAA** [comment]
谢谢！我仔细阅读一下

**Comment by NYB-AAA** [comment]
您好，请问这个模型可以训练自己的RGB格式数据集吗，具体操作步骤是什么呢，例如数据集的标注格式是什么

**Comment by xiaoqiang-lu** [comment]
你好，可以的，任何通用语义分割数据集都能够训练，只需更改数据集导入路径，语义类别数。标注格式就是单通道存有类别信息的掩码，.png，.tif等格式均可。

**Comment by NYB-AAA** [comment]
你好，你还记得你当时训练使用的seed设置是多少吗，我突然发现我一直用的默认1234seed🙈，因为突然发现测试集上复现结果差百分之1，在想是不是种子的问题，其他参数没有动过



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2023年12月6日(周三) 上午10:30
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [xiaoqiang-lu/LSST] 可以提供一下图片推断可视化的代码吗 (Issue #1)




 
你好，可以的，任何通用语义分割数据集都能够训练，只需更改数据集导入路径，语义类别数。标注格式就是单通道存有类别信息的掩码，.png，.tif等格式均可。

**Comment by xiaoqiang-lu** [comment]
具体种子我也忘记了。当初审稿人也问过不同种子的影响，我在每个数据集上实验了5种随机种子，最好的和最差的我记得好像能差距到百分之2，所以种子的影响也比较大，你可以选择“Torch.manual_seed(3407) is all you need”，但是我个人觉得波动大的主要原因是因为训练epoch只设置了50的原因。起初，出于公平比较的考虑，我设定了每种方法均进行100epoch，但是对于自训练的分阶段方法，我的做法是监督阶段50epoch，自训练阶段50epoch，论文中的结果就是这么跑的。但后来发现主流做法都是监督阶段和自训练阶段都跑100epoch，我也实验了下确实可以继续涨点。

**Comment by ghp0210** [comment]
您好，我使用自己的数据集，希望对遥感图像里的云做分割，但是运行不了
![image](https://github.com/xiaoqiang-lu/LSST/assets/128354814/ceca21b1-e5fa-48cb-9f78-da978e6e6cd8)

**Comment by xiaoqiang-lu** [comment]
这个问题出现的原因一般是semi.py文件在读取某些.tif文件时报错，一种做法是你将自己数据集改成.png格式，另一种做法就是将semi.py的第55行`img = Image.open(os.path.join(self.root, id.split(' ')[0]))`换成cv2读图`img = Image.fromarray(cv2.cvtColor(cv2.imread(os.path.join(self.root, id.split(' ')[0])), cv2.COLOR_BGR2RGB))`

**Comment by ghp0210** [comment]
您好，我准备做二分类（云和非云），运行结果一类IoU是100，另一类是nan，loss始终是0，是哪里出了问题吗
![image](https://github.com/xiaoqiang-lu/LSST/assets/128354814/e5318a4a-7242-43ff-8347-20ec27eb13ac)
我修改的dataset_name.py
![image](https://github.com/xiaoqiang-lu/LSST/assets/128354814/1a1db421-1856-460a-a538-86bc37cbd785)
![image](https://github.com/xiaoqiang-lu/LSST/assets/128354814/07b2884d-326f-4bf2-83eb-68cc1f41b287)
我的img和对应的label:
![image](https://github.com/xiaoqiang-lu/LSST/assets/128354814/779bcf79-caf5-45b3-b5ac-eb60d14167b1)
![image](https://github.com/xiaoqiang-lu/LSST/assets/128354814/2c9a457b-8336-407e-a2e0-6eb3a8915d39)
